### PR TITLE
Add textSelectionControls to TextField etc.

### PR DESCRIPTION
**Pull request by jamesblasco**
## Description

*Allows developers to add custom textSelectionControls to SelectableText.*

## Related Issues

https://github.com/flutter/flutter/issues/36985

## Tests

I added the following tests:

*Replace this with a list of the tests that you added as part of this PR. A change in behavior with no test covering it
will likely get reverted accidentally sooner or later. PRs must include tests for all changed/updated/fixed behaviors. See [Test Coverage].*

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I signed the [CLA].
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [X] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [X] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [X] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by justinmc** [comment]
@chunhtai Any reason that we shouldn't add this to SelectableText?

**Comment by justinmc** [comment]
@jamesblasco Could you write a quick test for this?

**Comment by jamesblasco** [comment]
Great, I will add it tomorrow. What do you think about adding the same param to TextField, CupertinoTextField, TextFormField?

**Comment by chunhtai** [comment]
@justinmc I probably missed it when i wrote the selectable text

**Comment by justinmc** [comment]
Ah I forgot that selectionControls is not already added to TextField et al, it's only on EditableText.

I'm going to hold off on merging this until we've planned out how to bring custom text selection menus to the framework.  I worry that we may expose this selectionControls param and then decide on a different API and have to remove it.

I'm working on that design doc now and will send it to you @jamesblasco.  Then we can decide whether to use selectionControls or not.

**Comment by justinmc** [comment]
@jamesblasco Could you add a test for this?  And could you also add this parameter to TextField, CupertinoTextField, and TextFormField and test those?

I think my plan with the text selection menu customization is to keep TextSelectionControls as-is, but clean up and publicly expose some of the classes that are currently private (similar to your project https://github.com/jamesblasco/selection_controls_example).  Then users can pass in whatever they want to selectionControls and use the public classes to build their toolbar.

I may deprecate canCopy/handleCopy etc., but I think it's ok to move forward with this PR now anyway.

I'm working on refactoring and cleaning up some of those internal classes now.

**Comment by jamesblasco** [comment]
I added the param and test to all of them. 

But now I just realized I think it would be better if the param is called `selectionControls` than `textSelectionControls`. Should I rename it?

File: packages/flutter/lib/src/cupertino/text_field.dart
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/text_field.dart:273]: As you mentioned, yes we should change this to `selectionControls` in order to be consistent with the parameter in EditableText.  Good catch!

File: packages/flutter/lib/src/cupertino/text_field.dart
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/text_field.dart:882]: Would it work to set a default value for the selectionControls parameter instead of doing this here?  What's your opinion on that?

File: packages/flutter/lib/src/material/selectable_text.dart
Comment (justinmc) [inline packages/flutter/lib/src/material/selectable_text.dart:574]: Could this be `late`?  I'm still getting the hang of the nnbd stuff.

File: packages/flutter/test/cupertino/text_field_test.dart
Comment (justinmc) [inline packages/flutter/test/cupertino/text_field_test.dart:51]: It looks like a code formatter was run in this file and it's hard to see what was actually changed, could you remove the formatting changes?

**Review by justinmc — COMMENTED** [review]
Thanks for updating this PR so quickly!  Looks good overall, just a few small changes, including the selectionControls name you mentioned.

File: packages/flutter/lib/src/material/selectable_text.dart
Comment (jamesblasco) [reply packages/flutter/lib/src/material/selectable_text.dart:574]: It wouldn't make any change. The late keyword allows postponing the initializing of a non-nullable variable. But in that line, it is being initialized already with a variable that could be null.

Notice that EditableText accepts null in the `selectionControls` param.

If non-nullable is prefered we could leave that line like before and inside the EditableText constructor use `widget.textSelectionControls ?? textSelectionControls`

File: packages/flutter/lib/src/cupertino/text_field.dart
Comment (jamesblasco) [reply packages/flutter/lib/src/cupertino/text_field.dart:882]: That is a great idea. The only thing is that `cupertinoTextSelectionControls` is a final variable and not const. We could create a const constructor inside `TextSelectionControls`

Comment (justinmc) [reply packages/flutter/lib/src/cupertino/text_field.dart]: Ah you're right.  I think it's fine like this then.

File: packages/flutter/test/cupertino/text_field_test.dart
Comment (justinmc) [inline packages/flutter/test/cupertino/text_field_test.dart:4173]: No need for this newline here.

File: packages/flutter/test/material/text_field_test.dart
Comment (justinmc) [inline packages/flutter/test/material/text_field_test.dart:6298]: Remove this extra newline here.

File: packages/flutter/test/material/text_field_test.dart
Comment (justinmc) [inline packages/flutter/test/material/text_field_test.dart:6299]: "select" => "selection"

File: packages/flutter/test/material/text_field_test.dart
Comment (justinmc) [inline packages/flutter/test/material/text_field_test.dart:6377]: Remove the extra newline here so that there is only one line between the two tests.

File: packages/flutter/test/widgets/selectable_text_test.dart
Comment (justinmc) [inline packages/flutter/test/widgets/selectable_text_test.dart:2776]: I think this test is indented too far compared to the other tests around it.

File: packages/flutter/test/widgets/selectable_text_test.dart
Comment (justinmc) [inline packages/flutter/test/widgets/selectable_text_test.dart:2844]: Some indentation is strange with this test too.

**Review by justinmc — COMMENTED** [review]
I think this looks great, thanks for making my changes.  Just a few comments about indentation and some small stuff like that.  Once that's fixed I think this is ready to merge.

**Comment by SeasonLeee** [comment]
@justinmc can I fix what you have pointed out and request for merge? 😅 it's three days ago... is there any progress on this? Anything I can help...? Thank you...

File: packages/flutter/test/material/text_field_test.dart
Comment (justinmc) [reply packages/flutter/test/material/text_field_test.dart:6376]: Bumping this comment in case you missed it.

**Comment by justinmc** [comment]
I think this is just about ready.  One comment above about an extra newline @jamesblasco (not a big deal but please remove the extra newline if you have time).  The Mac build tests are failing but I think it's an infrastructure problem, I'll try to get it to go green.

**Comment by jamesblasco** [comment]
Sorry about that @justinmc, I missed the newline. Let's see if the Mac build test works now. Also, the Linux gradle_r8_test has been running for 10h and not finished yet.

**Comment by justinmc** [comment]
@jamesblasco Thanks!  Could you merge master into this PR?  I think that's what we have to do to get that Linux gradle_r8_test to run again...

**Comment by SeasonLeee** [comment]
@justinmc hey, one quick question here, I have seen so many build fail because of the new line or wrong indentation, this code style kind of thing, while it can be annoying but can be great help when we want the codebase clean and readable, do you know what tool flutter is using here to enforce the code style? Thanks 😁

**Comment by jamesblasco** [comment]
I rebase the code to the latest master version. Let's see if this solves it.

@SeasonLeee The code styling for the Flutter repository is defined here https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#formatting. Most of the code style issues are detected by the dart analyzer and will make the build fail. A very few ones, like extra newlines, are not detected so it is important to make sure there is none before merging to master. 

While `dart format` can auto-format your code directly, it is not used in the Flutter repository, so you have to do it manually.

I am not sure this what you were asking but hope it can help 🙂

**Comment by SeasonLeee** [comment]
@jamesblasco oh! definitely!!! You explain how these things works pretty well, thank you so much! 

So, you just said that `dart format` didn't use in the Flutter repo, why is that?

No wonder there's a time I tried to check the code base and when I change something and, you know that Android Studio I set it to auto format, the whole file has been formatted! And there are not just no single thing I changed but a lot of changes because of the original file is not formatted!

It's quite strange, since I think they made the dart formatter thing...

Although we can still see that even after you tried something there but the build is still failing... 

image
 
I just wonder when will can I use this textSelectionControl thing😅 and how on earth does this keep failing.... 

By the way, I opened an issue regarding SelectableText, could you have a look on it? Thanks! 

https://github.com/flutter/flutter/issues/68675

I want to see the Flutter codebase on what cause that problem, but I lost in the pile of codes... god, how helpless am I 😭...

**Comment by jamesblasco** [comment]
Yess, auto-format is a problem with Flutter repository. Related to why Flutter doesn't use 'dart format' you can read in the link I send you: 

> We do not yet use dartfmt. Flutter code tends to use patterns that the standard Dart formatter does not handle well. We are working with Dart team to make dartfmt aware of these patterns.

Related to the flutter tests, `flutter-build` is a special test that indicates if the master branch is open to merges. It is not related to any PR. If it fails, it means you should wait until is green to merge it. (Notice it will auto-update itself). 

I will take a look at the issue you shared. Let's try to keep this PR focused on the issue, and feel free to ping me on Discord/Slack/Stackoverflow or email if you have more questions

**Comment by SeasonLeee** [comment]
@jamesblasco oh, man, you are right, now all the checks are green and it's passed!!!!! So now we are waiting for someone who have the permission to merge it into master and I `flutter upgrade`, then I will be able to use the selectionControl feature, right?! God, it's so exciting! And congratulation on your work!! And thank you for all the explanations!!

**Review by justinmc — APPROVED** [review]
LGTM 👍 

This is ready to merge, but there is some NNBD work going on so the tree is closed while that happens.  I will merge this as soon as that is done.

Update: Looks like that will be sometime tomorrow.

**Comment by jamesblasco** [comment]
Thank you for your time @justinmc!

**Comment by justinmc** [comment]
Thank you!

**Comment by SeasonLeee** [comment]
Thank you both for everything have done for this feature! I'm sure I will tag you when our product publish!!! Thank you!!! @jamesblasco @justinmc

**Comment by Hixie** [comment]
This may be responsible for a regression: https://github.com/flutter/flutter/issues/68890

**Pull request merged**
